implemented create

JamesSwandale opened the pull request.

JamesSwandale commented:
There are probably more elegant solutions to this, please give it a review and let me know any areas I can improve.

jesuspc commented on lib/dictionaries/api.js:
```diff
@@ -33,7 +33,21 @@ module.exports = function(opts){
 
   router.put('/:scope/:uuid/dictionaries/:name.json', parameterize, authorize, function(req, res, next) {
+    var filters = {'meta.uuid': req.params.uuid, 'meta.scope': req.params.scope, name: req.params.name};
+    if (!req.body.name){
```
We should move that logic into a parameterizer

jesuspc commented on lib/dictionaries/api.js:
```diff
@@ -33,7 +33,21 @@ module.exports = function(opts){
+      }
+    };
 
+    transactions.create(filters, payload).then(handleResponse);
```
It's weird to call a create action with filters. I think we should name it somehow in which the fact that an update or create is performed. Also payload is not a very nice name for the actual content of the dictionary. I would rather call it dictionaryData or simply dictionary.

jesuspc commented on lib/dictionaries/transactions/create.js:
```diff
@@ -1 +1,27 @@
+    var deferred = Q.defer();
+
+    dbConnection().then(function(db){
+      db.collection('dictionaries').findOneAndUpdate(filters, payload, {upsert: true, returnOriginal: false} ).then(function(dict){
```
There's an extra whitespace after the returnOriginal: false} :)

jesuspc commented on lib/dictionaries/transactions/create.js:
```diff
@@ -1 +1,27 @@
+      db.collection('dictionaries').findOneAndUpdate(filters, payload, {upsert: true, returnOriginal: false} ).then(function(dict){
+        deferred.resolve(dict.value);
+      })
+  	})
```
And a weird indentation here

jesuspc commented on lib/dictionaries/transactions/create.js:
```diff
@@ -1 +1,27 @@
+  	})
+    
+    return deferred.promise;
+  };;
```
Double semicolon

jesuspc commented on lib/dictionaries/transactions/create.js:
```diff
@@ -1 +1,27 @@
+};
+
+
+/*
```
Lets remove the comments. Git is supposed to be used to keep track of that kind of stuff

jesuspc commented on test/integration/createDictionary.spec.js:
```diff
@@ -4,42 +4,130 @@ var box = require('../box')(),
+        })
+
+        it('returns a 200', function(done){
+          that = this
```
That should be declared as var and should have a semicolon

jesuspc commented on test/integration/createDictionary.spec.js:
```diff
@@ -4,42 +4,130 @@ var box = require('../box')(),
+                db.collection('dictionaries').findOne({name: that.dictName}).then(function(dict){
+                    try{
+                      assert.deepEqual(dict.name, expectedPayload.name);
+                      assert.deepEqual(dict.unnamedField, expectedPayload.unnamedField);
```
Again, payload is not a very nice name

jesuspc commented on test/integration/createDictionary.spec.js:
```diff
@@ -4,42 +4,130 @@ var box = require('../box')(),
-
+        it('returns the created element as json', function(done){
+          var expectedBody = { name: "unnamedPayload", unnamedField: "unnamedValue" };
+          that = this;
```
var that

jesuspc commented on test/integration/createDictionary.spec.js:
```diff
@@ -4,42 +4,130 @@ var box = require('../box')(),
-
+      describe('when no payload provided', function(done){
+        it('returns a 200', function(done){
+          that = this
```
var thar, semicolon and indentation

jesuspc commented on test/integration/createDictionary.spec.js:
```diff
@@ -4,42 +4,130 @@ var box = require('../box')(),
+          that = this
+            this.doRequest(function(req){
+              req.expect(200, done);
+            })
```
A semicolon is missing here

jesuspc commented on test/integration/createDictionary.spec.js:
```diff
@@ -4,42 +4,130 @@ var box = require('../box')(),
+              dbConnection().then(function(db){
+                db.collection('dictionaries').findOne({name: that.dictName}).then(function(dict){
+                    try{
+                      assert.deepEqual(dict.name, expectedPayload.name);
```
Payload again

jesuspc commented:
Good work again! There are just some minor issues under my point of view. 

Btw, probably we should start unit testing stuff too

JamesSwandale commented:
Thanks for the comments, I've fixed the issues, I wasn't sure what to do with "move that logic into a parameterizer" currently I've moved it into a function to deal with later.
